### PR TITLE
More gl2d symbols!

### PR DESCRIPTION
**Pull request by etpinard**
Supersedes https://github.com/plotly/plotly.js/pull/1550 with better mocks and fine tuning via https://github.com/gl-vis/gl-scatter2d-sdf/pull/4 and https://github.com/plotly/plotly.js/commit/b1bccdc8993ff73175531939e432a690f998f885

TODO:

- [x] bump `gl-scatter2d-sdf` once @dfcreative 's https://github.com/gl-vis/gl-scatter2d-sdf/pull/4 is merged/published

### CODE REVIEW

File: test/image/mocks/marker_symbols.json [hunk 1]
```diff
@@ -0,0 +1,505 @@
+{
+  "data": [
```
Comment (etpinard) [inline test/image/mocks/marker_symbols.json]: generated with https://codepen.io/etpinard/pen/LLZGZV

**Comment by etpinard** [comment]
SVG:

marker_symbols


gl2d:

gl2d_marker_symbols

File: test/image/mocks/gl3d_marker_symbols.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      "text": [
+        "marker symbol: 'circle'",
+        "marker symbol: 'circle-open'",
+        "marker symbol: 'square'",
+        "marker symbol: 'square-open'",
+        "marker symbol: 'diamond'",
+        "marker symbol: 'diamond-open'",
+        "marker symbol: 'cross'"
+      ],
+      "marker": {
```
Comment (etpinard) [inline test/image/mocks/gl3d_marker_symbols.json]: We should make sure new marker symbols are positioned and scaled properly in gl3d before adding them. To be completed in a future PR.

**Comment by dfcreative** [comment]
@etpinard if marker set is going to be limited, a better solution would be pre-generated sdf sprite with all markers included. That would enable all svg markers in _scattergl_ and allow to get rid of _font-atlas-sdf_, which would make code of _gl-scatter2d_ way simpler and performant.
Probably it should be addressed in #949.

**Comment by alexcjohnson** [comment]
Could we make some of the markers be rotated characters? So like ➕  could rotate 45 degrees to make ✖️  rather than having to use letter X.

**Comment by dfcreative** [comment]
@alexcjohnson we use [font-atlas-sdf](https://github.com/dfcreative/font-atlas-sdf) to render symbols, which in turn uses [tiny-sdf](https://github.com/mourner/tiny-sdf) internally to draw characters. It is possible, but would unnecessarily bloat API. In fact it is not the X letter used, but valid [unicode box cross symbol](https://unicode-table.com/en/2573/) ╳, which is rendered so with default ubuntu sans-serif font. We can replace it with [any other cross](https://unicode-table.com/en/search/?q=cross).

**Comment by alexcjohnson** [comment]
OK - just trying to get closer to parity. Lots of good additions here, but I don't think we should include symbols that don't match fairly closely what we do in SVG. To my eye the red ones below are not close enough as shown, and the green ones could probably be used unfilled but not filled.
gl2d-markers

**Comment by etpinard** [comment]
@alexcjohnson 

> To my eye the red ones below are not close enough as shown, and the green ones could probably be used unfilled but not filled.

Done in :arrow_right: https://github.com/plotly/plotly.js/pull/1781/commits/bb9a46c808352503bca429abd4f74ef08c9fb455

Thanks for taking a look :microscope:

**Comment by dfcreative** [comment]
@alexcjohnson thanks for the good point! Fixed in 1f610febece7ec8b7bbee2cf50d7e08016248916

**Comment by etpinard** [comment]
> it's a bit of a shame not to have any "X" symbols...

Yeah, it's a shame. Actually, I should put [it](https://github.com/plotly/plotly.js/blob/85e8770998a0e0ca4ee3a3de3ede6f494a3ec91f/src/constants/gl_markers.js#L20) back in even it doesn't render in `imagetest`.

File: src/traces/scattergl/convert.js [hunk 3]
```diff
@@ -491,29 +500,56 @@ proto.updateFancy = function(options) {
...
+            } else {
+                _borderColors = borderColors;
+            }
+
+            if(symbolSpec.bwFactor) {
+                bwFactor = symbolSpec.bwFactor;
+            } else if(symbolSpec.noBorder) {
+                bwFactor = 0.25;
+            } else if(symbolSpec.noFill) {
+                bwFactor = 0.1;
```
Comment (alexcjohnson) [inline src/traces/scattergl/convert.js]: so... this means that for these symbols with a smaller `bwFactor` the border will grow more slowly than the others? The result being that for small `borderWidth` these ones be thicker than other symbols, and for large `borderWidth` they will be thinner, and somewhere in the middle they will look similar right? I would think we'd do better with something like a minimum border width, as a fraction of `size`, that corresponds to the line width already baked into the unicode symbol, so the symbol gets *no* explicit border until you provide a width greater than this, and at greater widths you set a border `borderWidth - minBorderWidth`.

**Comment by alexcjohnson** [comment]
@etpinard nice work! just the one more question about how we calculate border widths https://github.com/plotly/plotly.js/pull/1781/files#r121820798 and we should be ready to go. The available symbols are looking really good now. Thanks for bearing with all my nitpicks, I just want to get this all as final as possible so we don't need to come back to it again...

File: src/traces/scattergl/convert.js (hunk 3 above)
Comment (etpinard) [inline src/traces/scattergl/convert.js]: Thanks for the headsup. For documentation purposes, using https://codepen.io/etpinard/pen/JJRMxo?editors=0010 as of commit fffc702


SVG gives:

image

and gl:

image

Comment (etpinard) [inline src/traces/scattergl/convert.js]: @alexcjohnson here's my latest attempt:

https://github.com/plotly/plotly.js/compare/gl2d-more-symbols...gl2d-more-symbols-line-width

Comment (alexcjohnson) [inline src/traces/scattergl/convert.js]:
```
//this.scatter.options.borderWidths[i] = (bw > minBorderWidth) ? bw - minBorderWidth : 0;

// but I found better results with:
this.scatter.options.borderWidths[i] = (bw > minBorderWidth) ? bw - minBorderWidth : bw;
```

`minBorderWidth` should only be nonzero for `noBorder` or `noFill` symbols right? Does that make it work better to use `bw - minBorderWidth : 0`? As it is it looks funny, as border width is non-monotonic.

Comment (etpinard) [inline src/traces/scattergl/convert.js]: Yeah, I think that's better:

https://github.com/plotly/plotly.js/commit/3f664a1744467f191a230384dbf74549ab1452a2

Comment (alexcjohnson) [inline src/traces/scattergl/convert.js]: Awesome - I think we've got it 🎉  💃

**Comment by alexcjohnson** [comment]
💃 💃 💃 !

**Pull request merged**